### PR TITLE
Improve conditional forwarding settings

### PR DESCRIPTION
**Pull request by DL6ER**
**By submitting this pull request, I confirm the following:** 

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md), as well as this entire template.
- [X] I have made only one major change in my proposed changes.
- [X] I have commented my proposed changes within the code.
- [X] I have tested my proposed changes, and have included unit tests where possible.
- [X] I am willing to help maintain this change if there are issues with it later.
- [X] I give this submission freely and claim no ownership.
- [X] It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
- [X] I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

---
**What does this PR aim to accomplish?:**

"Conditional forwarding" can be used to tell FTL to send queries to local devices (either with a local domain) and PTR requests for private ranges to a dedicated device (typically the router of the network) instead of the configured upstreams.
While this works for most users, it is not very flexible, as it implies three severe limitations (assuming the router is `10.1.2.3`):
- It is limited to IPv4
- It automatically assumes a Class D network. i.e. `10.1.2.0/24`. This is often correct (typically for `192.168.x.0/24` networks, but may also be wrong!
- It requires the subnet to be the same Class D network the router is in (this may be a fair assumption, however).

**How does this PR accomplish the above?:**

This PR aims at making conditional forwarding more flexible.

- It supports IPv4 and IPv6
- It adds a dedicated box for the local subnet to be used (can be, e.g., `10.0.0.0/8` in above's example)
- The router can be specified independently of the subnet, it could even be a foreign device (like "ask `13.225.3.2` for all IPs in range `10.0.0.0/8`).

Screenshot from 2020-04-09 09-25-46

**This feature is intended for Pi-hole v5.1 to ensure we have proper time for testing once it hits `development`**

**What documentation changes (if any) are needed to support this PR?:**

It has to be checked whether/where we document conditional forwarding.

### CODE REVIEW

**Comment by LordFlashmeow** [comment]
You've given detailed instructions on how to set it up for IPv4, but nothing for IPv6. Is any documentation going to be added for that?

**Comment by DL6ER** [comment]
Hmmm, do we really need it? It is basically the same thing with the some exception of you using an IPv6 address.

**Comment by DL6ER** [comment]
Also: I submitted a patch some longer time ago to the dnsmasq mailing list which removed the limitations on the allowed CIDR size, however, no response to this as of yet.

**Comment by LordFlashmeow** [comment]
My only concern is that IPv6 is a bit of a mystery compared to IPv4 (at least for me). 

When I was looking for a way to set up conditional forwarding IPv6, all the discourse posts used a ULA in addition to the router's IP [like this](https://discourse.pi-hole.net/t/pi-hole-as-a-dhcp-dual-stack-ipv4-ipv6-and-clients-hostnames-fritzbox-settings-help/12347/5)

> Assume your ipv6 ULA-Prefix is `fd0a:1234:5678::/48`
> also assume `fd0a:1234:5678::1` is router’s ipv6 address
> 
> Add an entry into /etc/dnsmasq.d/01-pihole.conf as follows:
> 
> `server=/8.7.6.5.4.3.2.1.a.0.d.f.ip6.arpa/fd0a:1234:5678::1 `

Sorry if I'm missing something, but would the ULA be unneeded with this update?

**Comment by DL6ER** [comment]
@LordFlashmeow 
> Sorry if I'm missing something, but would the ULA be unneeded with this update?

No, you would simple enter

Screenshot from 2020-06-04 08-42-42

**Comment by LordFlashmeow** [comment]
Ok, in theory that makes sense. In practice, I guess we’ll see... :)

File: settings.php [hunk 1]
```diff
@@ -975,38 +976,69 @@ function convertseconds($argument)
...
+                                                    </div>
+                                                    <div class="input-group">
+                                                      <table class="table table-bordered">
+                                                        <tr>
+                                                          <th>Local network in <a href="https://en.wikipedia.org/wiki/Classless_Inter-Domain_Routing" target="_blank">CIDR notation</a></th>
+                                                          <th>IP address of your DHCP server (router)</th>
+                                                          <th>Local domain name (optional)</th>
+                                                        </tr>
+                                                        <tr>
+                                                          <div class="input-group">
```
Comment (XhmikosR) [inline settings.php]: This is invalid BTW, you can't have a `div` inside a `tr`

File: settings.php [hunk 2]
```diff
@@ -975,38 +976,69 @@ function convertseconds($argument)
...
+                                                <p>Validate DNS replies and cache DNSSEC data. When forwarding DNS
+                                                   queries, Pi-hole requests the DNSSEC records needed to validate
+                                                   the replies. If a domain fails validation or the upstream does not
+                                                   support DNSSEC, this setting can cause issues resolving domains.
+                                                   Use Google, Cloudflare, DNS.WATCH, Quad9, or another DNS
+                                                   server which supports DNSSEC when activating DNSSEC. Note that
+                                                   the size of your log might increase significantly
+                                                   when enabling DNSSEC. A DNSSEC resolver test can be found
+                                                   <a href="https://dnssec.vs.uni-due.de/" rel="noopener" target="_blank">here</a>.</p>
+                                                <label>Conditional forwarding</label>
```
Comment (XhmikosR) [inline settings.php]: Maybe using an `h3` or `h4` would be better. There's not a form here.

**Comment by XhmikosR** [comment]
I haven't tested this, just some general markup comments :)

File: settings.php (hunk 1 above)
Comment (DL6ER) [reply settings.php]: Thanks, addressed.

File: settings.php (hunk 2 above)
Comment (DL6ER) [reply settings.php]: Thanks, addressed.

File: settings.php [hunk 3]
```diff
@@ -975,38 +976,68 @@ function convertseconds($argument)
...
+                                                <p>Validate DNS replies and cache DNSSEC data. When forwarding DNS
+                                                   queries, Pi-hole requests the DNSSEC records needed to validate
+                                                   the replies. If a domain fails validation or the upstream does not
+                                                   support DNSSEC, this setting can cause issues resolving domains.
+                                                   Use Google, Cloudflare, DNS.WATCH, Quad9, or another DNS
+                                                   server which supports DNSSEC when activating DNSSEC. Note that
+                                                   the size of your log might increase significantly
+                                                   when enabling DNSSEC. A DNSSEC resolver test can be found
+                                                   <a href="https://dnssec.vs.uni-due.de/" rel="noopener" target="_blank">here</a>.</p>
+                                                <h4>Conditional forwarding</h4>
```
Comment (PromoFaux) [inline settings.php:989]: I think a line break here might be nicer, to break the content up a little.

Before:
image

```suggestion
                                                <br>
                                                <h4>Conditional forwarding</h4>
```

After:
image

**Review by PromoFaux — DISMISSED** [review]

**Review by PromoFaux — APPROVED** [review]

File: scripts/pi-hole/php/savesettings.php [hunk 4]
```diff
@@ -323,28 +347,35 @@ function addStaticDHCPLease($mac, $ip, $hostname) {
...
 				{
-					$conditionalForwardingIP = trim($_POST["conditionalForwardingIP"]);
-					$conditionalForwardingDomain = trim($_POST["conditionalForwardingDomain"]);
+					// Validate CIDR IP
+					$cidr = trim($_POST["rev_server_cidr"]);
+					if (!validCIDRIP($cidr))
+					{
+						$error .= "Conditional forwarding subnet (\"".htmlspecialchars($cidr)."\") is invalid!<br>".
+						          "This field requires CIDR notation for local subnets (e.g., 192.168.0.0/16).<br>".
+						          "Please use only subnets /8, /16, /24, and /32.<br>";
```
Comment (PromoFaux) [inline scripts/pi-hole/php/savesettings.php:359]: @DL6ER This works as expected, I get an error if I try to use a /23. But I do have a question, why is /23 invalid? 

My network spans 192.168.0.1-192.168.1.254

(I should have checked this before merging I guess!)

Comment (yubiuser) [reply scripts/pi-hole/php/savesettings.php:359]: https://discourse.pi-hole.net/t/support-prefix-12-for-conditonal-forwarding/33975?u=yubiuser

Comment (PromoFaux) [reply scripts/pi-hole/php/savesettings.php:359]: Thanks.... I tried a cursory search on discourse to see if it had already been discussed. I failed, clearly!

Comment (PromoFaux) [reply scripts/pi-hole/php/savesettings.php:359]: In any case, it works if I use /16

Comment (dschaper) [reply scripts/pi-hole/php/savesettings.php:359]: I think there was a mailinglist patch or question to allow for non-natural masks.

Comment (PromoFaux) [reply scripts/pi-hole/php/savesettings.php:359]: Yeah, for future reference:

http://lists.thekelleys.org.uk/pipermail/dnsmasq-discuss/2020q2/013985.html

http://lists.thekelleys.org.uk/pipermail/dnsmasq-discuss/2020q2/013997.html

Comment (DL6ER) [reply scripts/pi-hole/php/savesettings.php:359]: Yeah. TL;DR `dnsmasq` only supports multiples of 8 for IPv4 (8, 16, 24) and multiples of 4 for IPv6 (4, 8, 12, ...). I submitted patches to allow arbitrary prefix lengths but nothing happened to them. As odd prefix-lengths cause `dnsmasq` to fail on startup, they are forbidden for now (until my two patches are implemented)

Comment (PromoFaux) [reply scripts/pi-hole/php/savesettings.php:359]: > IPv5

Comment (DL6ER) [reply scripts/pi-hole/php/savesettings.php:359]: I do this every now and then, usually it goes unnoticed :-)

**Pull request merged**